HIP-584: Add tests for nested functions for call and estimate gas

anastasiya-kovaliova opened the pull request:
**Description:**

Introduces additional tests to validate the in-memory state via estimateGas and ethCall endpoints and nested functions. The new tests are executing newly added functions in the PrecompileTestContract contract. For each function there are tests for eth_call and estimate_gas for both fungible and NFT tokens.

* Add 9 contract nested functions
* Add 16 tests for call (ETHCALL-069 - ETHCALL-084)
* Add 16 tests for estimate gas (ETH_ESTIMATE-101 - ETH_ESTIMATE-116)

**Related issue(s):**

Fixes #7018

Notes for reviewer:

New PR. Old one is: https://github.com/hashgraph/hedera-mirror-node/pull/7085

Checklist

- [ ]  Documented (Code comments, README, etc.)
- [x]  Tested (unit, integration, etc.)

steven-sheehy commented:
Please address the code smells

steven-sheehy commented on hedera-mirror-test/src/test/resources/solidity/artifacts/@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol/IERC20Metadata.json:
Please remove any contracts or artifacts that get downloaded by npm and add a rule to gitignore to exclude them.

anastasiya-kovaliova replied:
Thank you! I've updated the gitignore file, now all contracts and artifacts downloaded by npm should be excluded.

steven-sheehy commented on hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/CallFeature.java:
```diff
@@ -84,23 +119,106 @@ public static String[] splitAddresses(String result) {
+        var nftStatus = mirrorClient.getTokenRelationships(receiverAccountId.getAccountId(), nonFungibleTokenId).getTokens().size();
+
+        //if the tokens are already associated from the parallel tests run skip the association
+        if(fungibleTokenStatus == 0){
```
nit: Should adjust your IDE save actions to auto format. Also, ensure the git hook is running to format files via spotless. Sometimes git UI tools don't respect the git hook.

```suggestion
        if (fungibleTokenStatus == 0) {
```

steven-sheehy commented on hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/CallFeature.java:
```diff
@@ -307,7 +698,18 @@ enum ContractMethods implements SelectorInterface {
         DEPLOY_NESTED_CONTRACT_CONTRACT_VIA_CREATE2_SELECTOR("deployNestedContracts2"),
-        REENTRANCY_CALL_WITH_GAS("reentrancyCallWithGas");
+        REENTRANCY_CALL_WITH_GAS("reentrancyCallWithGas"),
+        MINT_TOKEN_GET_TOTAL_SUPPLY_AND_BALANCE_OF_TREASURY_SELECTOR("mintTokenGetTotalSupplyAndBalanceOfTreasury"),
+        BURN_TOKEN_GET_TOTAL_SUPPLY_AND_BALANCE_OF_TREASURY_SELECTOR("burnTokenGetTotalSupplyAndBalanceOfTreasury"),
+        WIPE_TOKEN_GET_TOTAL_SUPPLY_AND_BALANCE_OF_TREASURY_SELECTOR("wipeTokenGetTotalSupplyAndBalanceOfAccount"),
+        PAUSE_GET_STATUS_UNPAUSE_GET_STATUS("pauseTokenGetPauseStatusUnpauseGetPauseStatus"),
+        FREEZE_TOKEN_GET_STATUS_UNFREEZE_GET_STATUS("freezeTokenGetFreezeStatusUnfreezeGetFreezeStatus"),
+        APPROVE_TOKEN_GET_ALLOWANCE("approveTokenGetAllowance"),
```
Some of my comments in the last PR about renaming methods/scenarios to be simpler did not see any change.

anastasiya-kovaliova replied:
Thank you for noticing that, I've updated them in all places

anastasiya-kovaliova commented on hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/util/TestUtil.java:
```diff
@@ -152,6 +152,18 @@ public static String getAbiFunctionAsJsonString(
+    }
+
+    public static String hexToAddress(String hex){
+        return Bytes.fromHexString(hex).slice(12).toUnprefixedHexString();
```
Yes, the address is 20 bytes, but the result comes with 12 bytes of 0s prefixed to the actual address, example value is: 0000000000000000000000007c17bae78b706463f9d8876352aca9536c52303e, after the slice the result is 7c17bae78b706463f9d8876352aca9536c52303e

steven-sheehy replied:
Think this assumes the address in question is a long zero address. Which maybe the current tests are but we shouldn't assume that will always be the case for a utility method. Think we should use `Bytes.fromHexString(hex).slice(20).trimLeadingZeros().toUnprefixedHexString()`. Or just change tests to verify the full zero prefixed string.

anastasiya-kovaliova replied:
That's great and won't narrow the function usage, thank you!

anastasiya-kovaliova replied:
Thank you! I've updated the test to verify the full zero prefixed string.

anastasiya-kovaliova commented on hedera-mirror-test/src/test/resources/features/contract/estimatePrecompile.feature:
```diff
@@ -158,6 +158,19 @@ Feature: EstimateGas Contract Base Coverage Feature
     Then I call estimateGas with exchange rate tinybars to tinycents
     #Then I call estimateGas with pseudo random seed
     #Then I call estimateGas with pseudo random number
+    Then I call estimate gas that mints FUNGIBLE token and returns the total supply and balance of treasury
+    Then I call estimate gas that mints NFT token and returns the total supply and balance of treasury
```
This is a great idea, I will add this to the QA roadmap and will create a ticket in the board so that we do not forget about it. Ticket number: #7144

anastasiya-kovaliova commented on hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/response/ContractCallResponse.java:
```diff
@@ -61,4 +64,22 @@ public String getResultAsAsciiString() {
         return hexToAscii(result.replace("0x", "").substring(128).trim());
     }
+
+    public List<Integer> getResultAsListDecimal() {
```
Thank you, I've updated the function to use Big Integer!

steven-sheehy commented:
Looks good, but getting this error with your branch when running against integration. Can you check?

```
    2023-11-02T22:13:29.861Z ERROR ForkJoinPool-1-worker-8 c.h.m.t.e.a.s.EstimatePrecompileFeature Scenario failure: 
    I create estimate precompile contract with 0 balance
    I create erc test contract with 0 balance
    I get exchange rates
    I successfully create Precompile contract with 0 balance
    I successfully create fungible tokens
    I successfully create non fungible tokens
    I mint and verify a new nft
    the mirror node REST API should return status 200 for the HAPI transaction
    I set lower deviation at 5% and upper deviation at 20%
    I call estimateGas with associate function for fungible token
    I call estimateGas with associate function for NFT
    I call estimateGas with dissociate token function without association for fungible token
    I call estimateGas with dissociate token function without association for NFT
    I call estimateGas with nested associate function that executes it twice for fungible token
    I call estimateGas with nested associate function that executes it twice for NFT
    I associate the receiver account with the fungible token
    the mirror node REST API should return status 200 for the HAPI transaction
    I call estimateGas with dissociate token function for fungible token
    I associate the receiver account with the NFT
    the mirror node REST API should return status 200 for the HAPI transaction
    I call estimateGas with dissociate token function for NFT
    I call estimateGas with dissociate and associate nested function for fungible token
    I call estimateGas with dissociate and associate nested function for NFT
    I call estimateGas with approve function without association
    I call estimateGas with approveNFT function without association
    I call estimateGas with setApprovalForAll function without association
    I associate contracts with the tokens and approve all nft serials
    the mirror node REST API should return status 200 for the HAPI transaction
    I call estimateGas with ERC approve function
    I call estimateGas with setApprovalForAll function
    I call estimateGas with transferFrom function without approval
    I call estimateGas with ERC transferFrom function without approval
    I approve receiver account to use the NFT with id 1 <- Transaction was unsuccessful: TransactionReceipt{transactionId=0.0.4661@1698963207.871000000, status=SENDER_DOES_NOT_OWN_NFT_SERIAL_NO, exchangeRate=ExchangeRate{hbars=1, cents=12, expirationTime=2100-01-01T00:00:00Z, exchangeRateInCents=12.0}, accountId=null, fileId=null, contractId=null, topicId=null, tokenId=null, topicSequenceNumber=null, topicRunningHash=null, totalSupply=0, scheduleId=null, scheduledTransactionId=null, serials=[], duplicates=[], children=[]}
    the mirror node REST API should return status 200 for the HAPI transaction
    I approve receiver account to use fungible token
    the mirror node REST API should return status 200 for the HAPI transaction
    I call estimateGas with ERC transferFrom function
    I call estimateGas with transferFrom function with more than the approved allowance
    I call estimateGas with ERC transferFrom function with more than the approved allowance
    I call estimateGas with transferFromNFT with invalid serial number
    I call estimateGas with transferNFT function
    I approve receiver account to use fungible token and transfer fungible token to the erc contract
    the mirror node REST API should return status 200 for the HAPI transaction
    I call estimateGas with ERC transfer function
    I call estimateGas with associateTokens function for fungible tokens
    I call estimateGas with associateTokens function for NFTs
    I associate the fungible_kyc_unfrozen token with the receiver account
    the mirror node REST API should return status 200 for the HAPI transaction
    I call estimateGas with dissociateTokens function for fungible tokens
    I associate the nft_kyc_unfrozen with the receiver account
    the mirror node REST API should return status 200 for the HAPI transaction
    I call estimateGas with dissociateTokens function for NFTs
    I associate and approve the second receiver to use the fungible_kyc_unfrozen token
    the mirror node REST API should return status 200 for the HAPI transaction
    I call estimateGas with transferTokens function
    I mint a new NFT and approve second receiver account to all serial numbers
    the mirror node REST API should return status 200 for the HAPI transaction
    I call estimateGas with transferNFTs function
    I call estimateGas with cryptoTransfer function for hbars
    I call estimateGas with cryptoTransfer function for nft
    I call estimateGas with cryptoTransfer function for fungible tokens
    I call estimateGas with burnToken function for fungible token
    I call estimateGas with burnToken function for NFT
    I call estimateGas with CreateFungibleToken function
    I call estimateGas with CreateNFT function
    I call estimateGas with CreateFungibleToken function with custom fees
    I call estimateGas with CreateNFT function with custom fees
    I approve and transfer fungible tokens to receiver account
    the mirror node REST API should return status 200 for the HAPI transaction
    I call estimateGas with WipeTokenAccount function
    I call estimateGas with WipeTokenAccount function with invalid amount
    I transfer NFT to receiver account
    the mirror node REST API should return status 200 for the HAPI transaction
    I call estimateGas with WipeNFTAccount function
    I call estimateGas with WipeNFTAccount function with invalid serial number
    I call estimateGas with GrantKYC function for fungible token
    I call estimateGas with GrantKYC function for NFT
    I call estimateGas with RevokeTokenKYC function for fungible token
    I call estimateGas with RevokeTokenKYC function for NFT
    I call estimateGas with Grant and Revoke KYC nested function
    I call estimateGas with Freeze function for fungible token
    I call estimateGas with Freeze function for NFT
    I call estimateGas with Unfreeze function for fungible token
    I call estimateGas with Unfreeze function for NFT
    I call estimateGas with nested Freeze and Unfreeze function for fungible token
    I call estimateGas with nested Freeze and Unfreeze function for NFT
    I call estimateGas with delete function for Fungible token
    I call estimateGas with delete function for NFT
    I call estimateGas with delete function for invalid token address
    I call estimateGas with updateTokenExpiryInfo function
    I call estimateGas with updateTokenInfo function
    I call estimateGas with updateTokenKeys function
    I call estimateGas with pause function for fungible token
    I call estimateGas with pause function for NFT
    I call estimateGas with unpause function for fungible token
    I call estimateGas with unpause function for NFT
    I call estimateGas for nested pause and unpause function
    I call estimateGas for nested pause, unpause NFT function
    I call estimateGas with getTokenExpiryInfo function
    I call estimateGas with isToken function
    I call estimateGas with getTokenKey function for supply
    I call estimateGas with getTokenKey function for KYC
    I call estimateGas with getTokenKey function for freeze
    I call estimateGas with getTokenKey function for admin
    I call estimateGas with getTokenKey function for wipe
    I call estimateGas with getTokenKey function for fee
    I call estimateGas with getTokenKey function for pause
    I call estimateGas with ERC allowance function for fungible token
    I call estimateGas with getApproved function for NFT
    I call estimateGas with ERC getApproved function for NFT
    I call estimateGas with isApprovedForAll function
    I call estimateGas with ERC isApprovedForAll function
    I call estimateGas with name function for fungible token
    I call estimateGas with name function for NFT
    I call estimateGas with symbol function for fungible token
    I call estimateGas with symbol function for NFT
    I call estimateGas with decimals function for fungible token
    I call estimateGas with totalSupply function for fungible token
    I call estimateGas with totalSupply function for NFT
    I call estimateGas with ownerOf function for NFT
    I call estimateGas with tokenURI function for NFT
    I call estimateGas with getFungibleTokenInfo function
    I call estimateGas with getNonFungibleTokenInfo function
    I call estimateGas with getTokenInfo function for fungible
    I call estimateGas with getTokenInfo function for NFT
    I call estimateGas with getTokenDefaultFreezeStatus function for fungible token
    I call estimateGas with getTokenDefaultFreezeStatus function for NFT
    I call estimateGas with getTokenDefaultKycStatus function for fungible token
    I call estimateGas with getTokenDefaultKycStatus function for NFT
    I call estimateGas with isKyc function for fungible token
    I call estimateGas with isKyc function for NFT
    I call estimateGas with isFrozen function for fungible token
    I call estimateGas with isFrozen function for NFT
    I call estimateGas with getTokenType function for fungible token
    I call estimateGas with getTokenType function for NFT
    I call estimateGas with redirect balanceOf function
    I call estimateGas with redirect name function
    I call estimateGas with redirect name function for NFT
    I call estimateGas with redirect symbol function
    I call estimateGas with redirect symbol function for NFT
    I call estimateGas with redirect decimals function
    I call estimateGas with redirect allowance function
    I call estimateGas with redirect getOwnerOf function
    I call estimateGas with redirect tokenURI function
    I call estimateGas with redirect isApprovedForAll function
    I transfer fungible token to the precompile contract
    the mirror node REST API should return status 200 for the HAPI transaction
    I call estimateGas with redirect transfer function
    I call estimateGas with redirect transferFrom function
    I call estimateGas with redirect approve function
    I call estimateGas with redirect transferFrom NFT function
    I call estimateGas with redirect setApprovalForAll function
    I call estimateGas with exchange rate tinycents to tinybars
    I call estimateGas with exchange rate tinybars to tinycents
    I call estimate gas that mints FUNGIBLE token and gets the total supply and balance
    I call estimate gas that mints NFT token and gets the total supply and balance
    I call estimate gas that burns FUNGIBLE token and gets the total supply and balance
    I call estimate gas that wipes FUNGIBLE token and gets the total supply and balance
    I call estimate gas that wipes NFT token and gets the total supply and balance
    I call estimate gas that pauses FUNGIBLE token, unpauses and gets the token status
    I call estimate gas that pauses NFT token, unpauses and gets the token status
    I call estimate gas that freezes FUNGIBLE token, unfreezes and gets freeze status
    I call estimate gas that freezes NFT token, unfreezes and gets freeze status
    I call estimate gas that approves FUNGIBLE token and gets allowance
    I call estimate gas that approves NFT token and gets allowance
    I call estimateGas with balanceOf function for "FUNGIBLE" and verify the estimated gas against HAPI
    I call estimateGas with balanceOf function for "NFT" and verify the estimated gas against HAPI
    I update the account and token keys
    the mirror node REST API should return status 200 for the HAPI transaction
    I call estimateGas with transferToken function and verify the estimated gas against HAPI
    I associate the contract with the receiver account
    the mirror node REST API should return status 200 for the HAPI transaction
    I call estimateGas with transferNFT function and verify the estimated gas against HAPI
    I call estimateGas with mintToken function for fungible token and verify the estimated gas against HAPI
    I call estimateGas with mintToken function for NFT and verify the estimated gas against HAPI
    I approve the receiver to use the token
    the mirror node REST API should return status 200 for the HAPI transaction
    I call estimateGas with allowance function for fungible token and verify the estimated gas against HAPI
    I call estimateGas with allowance function for NFT and verify the estimated gas against HAPI
    I call estimateGas with approve function and verify the estimated gas against HAPI
    I call estimateGas with approveNFT function and verify the estimated gas against HAPI
    I call estimateGas with transferFrom function with fungible and verify the estimated gas against HAPI
    I call estimateGas with transferFromNFT function and verify the estimated gas against HAPI
    I call estimate gas that approves a FUNGIBLE token and transfers it
    I approve and transfer NFT tokens to the precompile contract
    the mirror node REST API should return status 200 for the HAPI transaction
    I call estimate gas that approves a NFT token and transfers it
     com.hedera.mirror.test.e2e.acceptance.client.NetworkException: Transaction was unsuccessful: TransactionReceipt{transactionId=0.0.4661@1698963207.871000000, status=SENDER_DOES_NOT_OWN_NFT_SERIAL_NO, exchangeRate=ExchangeRate{hbars=1, cents=12, expirationTime=2100-01-01T00:00:00Z, exchangeRateInCents=12.0}, accountId=null, fileId=null, contractId=null, topicId=null, tokenId=null, topicSequenceNumber=null, topicRunningHash=null, totalSupply=0, scheduleId=null, scheduledTransactionId=null, serials=[], duplicates=[], children=[]}
        at com.hedera.mirror.test.e2e.acceptance.client.AbstractNetworkClient.getTransactionReceipt(AbstractNetworkClient.java:152)
        at com.hedera.mirror.test.e2e.acceptance.client.AbstractNetworkClient.executeTransactionAndRetrieveReceipt(AbstractNetworkClient.java:123)
        at com.hedera.mirror.test.e2e.acceptance.client.AbstractNetworkClient.executeTransactionAndRetrieveReceipt(AbstractNetworkClient.java:140)
        at com.hedera.mirror.test.e2e.acceptance.client.AccountClient.approveNft(AccountClient.java:254)
        at com.hedera.mirror.test.e2e.acceptance.steps.EstimatePrecompileFeature.approveNonFungibleWithReceiver(EstimatePrecompileFeature.java:514)
        at ✽.I approve receiver account to use the NFT with id 1(classpath:features/contract/estimatePrecompile.feature:37)

```

steven-sheehy commented:
There's a compilation error and my previous comment is not addressed.

anastasiya-kovaliova commented:
@steven-sheehy  still debugging and working on the proper way of resolving the failing tests during parallel run, I've updated the PR with the resolved merge conflicts and now the build should be passing and the parallel tests will be still failing with in the same place

anastasiya-kovaliova commented:
@steven-sheehy I've updated the tests to run in parallel successfully - changed one the of the used accounts in the eth_call tests and added new NFT token because with only 1 NFT the tests were conflicting every time in different places. Also removed 2 unused variables and updated and uncommented 2 checks for transfer after dissociate that is now fixed.

steven-sheehy commented:
It looks better, but still see a couple failures against integration. Running acceptance tests from main pass. Can you check?
```
    the mirror node REST API should return status 200 for the HAPI transaction
    I call estimateGas with transferToken function and verify the estimated gas against HAPI
    I associate the contract with the receiver account
    the mirror node REST API should return status 200 for the HAPI transaction
    I call estimateGas with transferNFT function and verify the estimated gas against HAPI
    I call estimateGas with mintToken function for fungible token and verify the estimated gas against HAPI
    I call estimateGas with mintToken function for NFT and verify the estimated gas against HAPI
    I approve the receiver to use the token
    the mirror node REST API should return status 200 for the HAPI transaction
    I call estimateGas with allowance function for fungible token and verify the estimated gas against HAPI
    I call estimateGas with allowance function for NFT and verify the estimated gas against HAPI
    I call estimateGas with approve function and verify the estimated gas against HAPI
    I call estimateGas with approveNFT function and verify the estimated gas against HAPI
    I call estimateGas with transferFrom function with fungible and verify the estimated gas against HAPI
    I call estimateGas with transferFromNFT function and verify the estimated gas against HAPI
    I call estimate gas that approves a FUNGIBLE token and transfers it <- 400 Bad Request from POST /api/v1/contracts/call
    I approve and transfer NFT tokens to the precompile contract
    the mirror node REST API should return status 200 for the HAPI transaction
    I call estimate gas that approves a NFT token and transfers it
     org.springframework.web.reactive.function.client.WebClientResponseException$BadRequest: 400 Bad Request from POST /api/v1/contracts/call
        at org.springframework.web.reactive.function.client.WebClientResponseException.create(WebClientResponseException.java:307)
        at ✽.I call estimate gas that approves a FUNGIBLE token and transfers it(classpath:features/contract/estimatePrecompile.feature:195)
        Suppressed: reactor.core.publisher.FluxOnAssembly$OnAssemblyException: 
    Error has been observed at the following site(s):
        *__checkpoint ⇢ 400 BAD_REQUEST from POST /api/v1/contracts/call [DefaultWebClient]
    Original Stack Trace:
```

And
```
    I mint a NFT
    the mirror node should return status 200 for the HAPI transaction
    I approve and transfer NFT token to the precompile contract
    the mirror node should return status 200 for the HAPI transaction
    I approve a FUNGIBLE token and get allowance
    I approve a NFT token and get allowance
    I dissociate a FUNGIBLE token and fail transfer
    I dissociate a NFT and fail transfer
    I approve a FUNGIBLE token and transfer it <- 400 Bad Request from POST /api/v1/contracts/call
    I approve a NFT token and transfer it
    I associate FUNGIBLE_KYC_UNFROZEN token to receiver account
    the mirror node should return status 200 for the HAPI transaction
    I grant and revoke KYC
    I successfully update the balance of an account and get the updated balance
     org.springframework.web.reactive.function.client.WebClientResponseException$BadRequest: 400 Bad Request from POST /api/v1/contracts/call
        at org.springframework.web.reactive.function.client.WebClientResponseException.create(WebClientResponseException.java:307)
        at ✽.I approve a FUNGIBLE token and transfer it(classpath:features/contract/call.feature:57)
        Suppressed: reactor.core.publisher.FluxOnAssembly$OnAssemblyException: 
    Error has been observed at the following site(s):
        *__checkpoint ⇢ 400 BAD_REQUEST from POST /api/v1/contracts/call [DefaultWebClient]
    Original Stack Trace:
```

anastasiya-kovaliova commented:
@steven-sheehy just updated the tests. After thе fix for the proper approve workflow has been merged there was a need to update these acceptance tests as well - the solidity functions "approveFungibleTokenTransferFromGetAllowanceGetBalance" and "approveNftAndTransfer" were updated to follow the proper workflow of transferring tokens from a contract to EOA in a nested function.

0xivanov commented on hedera-mirror-test/src/test/resources/solidity/contracts/PrecompileTestContract.sol:
```diff
@@ -277,4 +277,299 @@ contract PrecompileTestContract is HederaTokenService {
+        responseCode = HederaTokenService.revokeTokenKyc(token, account);
+        handleResponseCode(responseCode, "Revoke kyc failed.");
+
+        transferStatusAfterRevoke = HederaTokenService.transferToken(token, from, account, int64(uint64(amount)));
```
I've ran this test and for me it fails with com.hedera.node.app.service.evm.exceptions.InvalidTransactionException: TOKEN_NOT_ASSOCIATED_TO_ACCOUNT
so we are not hitting our required case (we would want it to fail because of the kyc)

0xivanov replied:
You should add association for precompileContract to the kycToken

0xivanov replied:
Please check other negative cases for the same issue if there are more.
I'm not sure if we can check the error programmatically here...
@steven-sheehy @IvanKavaldzhiev

0xivanov replied:
NVM, didn't see that you check the status and my setup was invalid so in failed with different status for me

yiliev0 reviewed (approved):
LGTM

anastasiya-kovaliova commented on hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/CallFeature.java:
```diff
@@ -308,12 +391,321 @@ public void ethCallReentrancyCallFunction() {
+        assertThat(results).isNotNull().hasSize(4);
+
+        // BalanceBefore + amount = balanceAfter
+        assertThat(intValue(intValue(results.get(0)) + 1)).isEqualTo(intValue(results.get(1)));
```
it is not, good catch!

0xivanov reviewed (dismissed):
LGTM, needs update from main.

steven-sheehy commented on hedera-mirror-test/src/test/resources/features/contract/estimatePrecompile.feature:
```diff
@@ -156,6 +156,21 @@ Feature: EstimateGas Contract Base Coverage Feature
     Then I call estimateGas with redirect setApprovalForAll function
     Then I call estimateGas with exchange rate tinycents to tinybars
     Then I call estimateGas with exchange rate tinybars to tinycents
+    #Then I call estimateGas with pseudo random seed
+    #Then I call estimateGas with pseudo random number
```
These commented out lines duplicate commented out lines below.

anastasiya-kovaliova replied:
Thank you for noticing that! I've removed them.

steven-sheehy commented on hedera-mirror-test/src/test/resources/features/contract/estimatePrecompile.feature:
```diff
@@ -156,6 +156,21 @@ Feature: EstimateGas Contract Base Coverage Feature
+    Then I call estimate gas that freezes NFT token, unfreezes and gets freeze status
+    Then I call estimate gas that approves FUNGIBLE token and gets allowance
+    Then I call estimate gas that approves NFT token and gets allowance
+    #Then I call estimate gas that associates FUNGIBLE token dissociates and fails token transfer
+    #Then I call estimate gas that associates NFT token dissociates and fails token transfer
```
Why are these commented out? If they're blocked on a ticket being fixed please add a comment to the file with the bug number

anastasiya-kovaliova replied:
Uncommented and updated the relevant contract functions

steven-sheehy commented on hedera-mirror-test/src/test/resources/features/contract/call.feature:
```diff
@@ -11,20 +11,55 @@ Feature: eth_call Contract Base Coverage Feature
-    Then I call function with HederaTokenService getTokenDefaultFreezeStatus token <fungible>
-    Then I call function with HederaTokenService getTokenDefaultKycStatus token <fungible>
+    Then I call function with IERC721Metadata token NFT totalSupply
+    Then I call function with IERC721 token NFT balanceOf owner
+    Then I call function with HederaTokenService isToken token FUNGIBLE
```
Some lines use `<nft>`/`<fungible>` and some use NFT/FUNGIBLE. It should be consistent. If we remove `<>` everywhere then we should delete the examples section.

anastasiya-kovaliova replied:
Removed <> everywhere and the Examples section as well, could be a good idea to refactor the whole suite in a proper way in a seperate task

steven-sheehy commented on hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/TokenClient.java:
```diff
@@ -274,32 +276,56 @@ public NetworkTransactionResponse createNonFungibleToken(
+                return response;
+            });
+        }
+        log.info("Token {} already associated to account {}", token, accountId);
```
Change to debug

steven-sheehy commented on hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/TokenClient.java:
```diff
@@ -274,32 +276,56 @@ public NetworkTransactionResponse createNonFungibleToken(
+        var account = new EntityId(accountId);
+        TokenAccountPair key = new TokenAccountPair(token, account);
+
+        synchronized (associations) {
+            associations.computeIfAbsent(key, k -> {
```
Since you're using a concurrent hash map synchronized is not necessary. Also, separate call to get() is not needed either:
```suggestion
        return associations.computeIfAbsent(key, k -> {
```

anastasiya-kovaliova replied:
Great idea! Now it is updated

steven-sheehy commented on hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/TokenClient.java:
```diff
@@ -274,32 +276,56 @@ public NetworkTransactionResponse createNonFungibleToken(
-        log.info("Associated account {} with token {} via {}", accountId, token, response.getTransactionId());
-        tokenAccounts.add(new TokenAccount(token, accountId));
-        return response;
+        var account = new EntityId(accountId);
+        TokenAccountPair key = new TokenAccountPair(token, account);
```
I don't think we should have a second object that holds token plus account nor the EntityId object. Why can't you just do:
```suggestion
        var key = new TokenAccount(token, accountId);
```

anastasiya-kovaliova replied:
Great idea! Now it is updated

steven-sheehy commented on hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/TokenClient.java:
```diff
@@ -274,32 +276,56 @@ public NetworkTransactionResponse createNonFungibleToken(
     public NetworkTransactionResponse associate(ContractId contractId, TokenId token)
             throws InvalidProtocolBufferException {
-        TokenAssociateTransaction tokenAssociateTransaction = new TokenAssociateTransaction()
-                .setAccountId(AccountId.fromBytes(contractId.toBytes()))
```
We should be able to combine associate account and associate contract into one method by convert ContractId to an ExpandedAccountId and calling `associate(ExpandedAccountId, TokenId)`.

anastasiya-kovaliova replied:
Great idea! Now there is only 1 method

steven-sheehy commented on hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/TokenClient.java:
```diff
@@ -66,6 +66,8 @@ public class TokenClient extends AbstractNetworkClient {
     private final Collection<TokenId> tokenIds = new CopyOnWriteArrayList<>();
 
     private final Map<TokenNameEnum, TokenResponse> tokenMap = new ConcurrentHashMap<>();
+    private final ConcurrentHashMap<TokenAccountPair, NetworkTransactionResponse> associations =
+            new ConcurrentHashMap<>();
```
We should remove tokenAccounts list since it's now redundant and use this in its place.

anastasiya-kovaliova replied:
Great idea! Now it is updated

steven-sheehy commented on hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/TokenClient.java:
```diff
@@ -274,32 +276,56 @@ public NetworkTransactionResponse createNonFungibleToken(
+        TokenAccountPair key = new TokenAccountPair(token, account);
+
+        synchronized (associations) {
+            associations.computeIfAbsent(key, k -> {
```
Ensure associations map is cleaned up during dissociate()

anastasiya-kovaliova replied:
Updated the dissociate as well

steven-sheehy commented on hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/CallFeature.java:
```diff
@@ -312,12 +396,321 @@ public void ethCallReentrancyCallFunction() {
+
+        assertThat(results).isNotNull().hasSize(4);
+
+        // allowed address before transfer should be the receiverAccount
+        assertThat(results.get(0)).isEqualTo(results.get(3));
```
nit: Comments like these would be more useful as assertj descriptions so they show up in test failure messages.
```suggestion
        assertThat(results.get(0)).as("allowed address before transfer should be the receiverAccount").isEqualTo(results.get(3));
```

anastasiya-kovaliova replied:
Great idea! Updated all tests with these kind of comments as suggested and it looks way better

steven-sheehy reviewed (changes requested):
Ran this against previewnet which has 0.93.0-beta1 and it failed:

```
    2023-11-23T16:13:16.124Z ERROR reactor-http-nio-6 c.h.m.t.e.a.c.MirrorNodeClient Endpoint failed, returning: 400 Bad Request from POST https://previewnet.mirrornode.hedera.com/api/v1/contracts/call 
    2023-11-23T16:13:16.126Z ERROR ForkJoinPool-1-worker-8 c.h.m.t.e.a.s.EstimatePrecompileFeature Scenario failure: 
    I create estimate precompile contract with 0 balance
    I create erc test contract with 0 balance
    I get exchange rates
    I successfully create Precompile contract with 0 balance
    I successfully create fungible tokens
    I successfully create non fungible tokens
    I mint and verify a new nft
    the mirror node REST API should return status 200 for the HAPI transaction
    I set lower deviation at 5% and upper deviation at 20%
    I call estimateGas with associate function for fungible token
    I call estimateGas with associate function for NFT
    I call estimateGas with dissociate token function without association for fungible token
    I call estimateGas with dissociate token function without association for NFT
    I call estimateGas with nested associate function that executes it twice for fungible token
    I call estimateGas with nested associate function that executes it twice for NFT
    I associate the receiver account with the fungible token
    the mirror node REST API should return status 200 for the HAPI transaction
    I call estimateGas with dissociate token function for fungible token
    I associate the receiver account with the NFT
    the mirror node REST API should return status 200 for the HAPI transaction
    I call estimateGas with dissociate token function for NFT
    I call estimateGas with dissociate and associate nested function for fungible token
    I call estimateGas with dissociate and associate nested function for NFT
    I call estimateGas with approve function without association
    I call estimateGas with approveNFT function without association
    I call estimateGas with setApprovalForAll function without association
    I associate contracts with the tokens and approve all nft serials
    the mirror node REST API should return status 200 for the HAPI transaction
    I call estimateGas with ERC approve function
    I call estimateGas with setApprovalForAll function
    I call estimateGas with transferFrom function without approval
    I call estimateGas with ERC transferFrom function without approval
    I approve receiver account to use the NFT with id 1
    the mirror node REST API should return status 200 for the HAPI transaction
    I approve the contract to use fungible token
    the mirror node REST API should return status 200 for the HAPI transaction
    I call estimateGas with transferFrom function with more than the approved allowance
    I call estimateGas with ERC transferFrom function with more than the approved allowance
    I call estimateGas with transferFromNFT with invalid serial number
    I call estimateGas with transferNFT function
    I approve the receiver account to use fungible token and transfer fungible token to the erc contract
    the mirror node REST API should return status 200 for the HAPI transaction
    I call estimateGas with ERC transferFrom function
    I call estimateGas with ERC transfer function
    I call estimateGas with associateTokens function for fungible tokens <- 400 Bad Request from POST https://previewnet.mirrornode.hedera.com/api/v1/contracts/call
    I call estimateGas with associateTokens function for NFTs
```

steven-sheehy commented on hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/TokenClient.java:
```diff
@@ -74,9 +73,10 @@ public TokenClient(SDKClient sdkClient, RetryTemplate retryTemplate) {
         deleteAll(tokenIds, tokenId -> delete(admin, tokenId));
-        deleteAll(tokenAccounts, tokenAccount -> dissociate(tokenAccount.accountId, tokenAccount.tokenId));
+        deleteAll(associations.keySet(), association -> dissociate(association.accountId, association.tokenId));
+        associations.clear();
```
nit: It's redundant to clear the associations since the deleteAll will already do so.

anastasiya-kovaliova replied:
Great catch, I removed it, thank you!

steven-sheehy reviewed (dismissed):
LGTM. It mostly passes against previewnet. I see the below error but I've also seen that error with main so don't believe it's related to your changes.

```
    I call the erc contract via the mirror node REST API for token symbol
    I call the erc contract via the mirror node REST API for token decimals
    I call the erc contract via the mirror node REST API for token totalSupply
    I call the erc contract via the mirror node REST API for token ownerOf <- 
    expected: "000000000000000000000000000000000001dd1f"
     but was: "0000000000000000000000000000000000000000"
    I call the erc contract via the mirror node REST API for token tokenUri
    I call the erc contract via the mirror node REST API for token getApproved
    I call the erc contract via the mirror node REST API for token allowance
    I call the erc contract via the mirror node REST API for token isApprovedForAll
    I call the erc contract via the mirror node REST API for token balance
    I approve "BOB" for nft
    the mirror node REST API should return status 200 for the erc contract transaction
    I call the erc contract via the mirror node REST API for token getApproved with response BOB
    I approve "ALICE" for nft all serials
    the mirror node REST API should return status 200 for the erc contract transaction
    I call the erc contract via the mirror node REST API for token isApprovedForAll with response true
    I approve "ALICE" with 2
    the mirror node REST API should return status 200 for the erc contract transaction
    I call the erc contract via the mirror node REST API for token allowance with allowances
    I call the erc contract via the mirror node REST API for token isApprovedForAll with response true with alias accounts
    I call the erc contract via the mirror node REST API for token allowance with alias accounts
    I call the erc contract via the mirror node REST API for token balance with alias account
     org.opentest4j.AssertionFailedError: 
    expected: "000000000000000000000000000000000001dd1f"
     but was: "0000000000000000000000000000000000000000"
        at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
        at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
        at com.hedera.mirror.test.e2e.acceptance.client.AbstractNetworkClient.validateAddress(AbstractNetworkClient.java:202)
        at com.hedera.mirror.test.e2e.acceptance.steps.ERCContractFeature.ownerOfContractCall(ERCContractFeature.java:130)
        at ✽.I call the erc contract via the mirror node REST API for token ownerOf(classpath:features/contract/erc.feature:14)
```

steven-sheehy commented:
Not all commits are GPG signed. Probably because you used GitHub to merge from main.

anastasiya-kovaliova commented:
Closing this PR and creating a clean one from another branch will all commits signed: https://github.com/hashgraph/hedera-mirror-node/pull/7383
